src/interface/swarm_boundaries.hpp

```diff
@@ -0,0 +1,139 @@
+//========================================================================================
+// (C) (or copyright) 2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SWARM_BOUNDARIES_HPP_
+#define INTERFACE_SWARM_BOUNDARIES_HPP_
+
+#include "swarm_device_context.hpp"
+
+namespace parthenon {
+
+class ParticleBound {
+ public:
+  KOKKOS_INLINE_FUNCTION virtual void Apply(const int n, double &x, double &y, double &z,
+                                            // const SwarmDeviceContext &context) = 0;
+                                            const SwarmDeviceContext &context) const {
+    PARTHENON_FAIL("Calling base class");
+  }
+};
+
+class ParticleBoundIX1Periodic : public ParticleBound {
+ public:
+  KOKKOS_INLINE_FUNCTION void Apply(const int n, double &x, double &y, double &z,
+                                    const SwarmDeviceContext &swarm_d) const override {
+    if (x < swarm_d.x_min_global_) {
+      x = swarm_d.x_max_global_ - (swarm_d.x_min_global_ - x);
+    }
+  }
+};
+
+class ParticleBoundIX1Outflow : public ParticleBound {
+ public:
+  KOKKOS_INLINE_FUNCTION void Apply(const int n, double &x, double &y, double &z,
+                                    const SwarmDeviceContext &swarm_d) const override {
+    swarm_d.MarkParticleForRemoval(n);
```

Comment:
Why is there no check required if the particle actually crossed the boundary?
If the function is only called if a particle already crossed it, then why is the check required for periodic boundary conditions?

Reply:
Yes good catch this logic is wrong, I need to add the checks. This will also be exercised with a modified `particle_leapfrog` regression test where only *some* of the particles should cross the boundaries and be removed.

Reply:
Checks added, regression test added